File: historyscan-back/src/main/java/io/sch/historyscan/infrastructure/features/analysis/history/HistoryMapper.java
package io.sch.historyscan.infrastructure.features.analysis.history;

import io.sch.historyscan.domain.contexts.analysis.common.CodeBaseCommit;
import io.sch.historyscan.domain.contexts.analysis.history.CodeBaseHistory;
import io.sch.historyscan.domain.contexts.analysis.history.CodeBaseHistoryCommitFile;
import io.sch.historyscan.domain.contexts.analysis.history.CodeBaseHistoryCommitInfo;
import io.sch.historyscan.infrastructure.features.analysis.history.CodeBaseHistoryCommitFileDTO;
import io.sch.historyscan.infrastructure.features.analysis.history.CodeBaseHistoryCommitInfoDTO;
import io.sch.historyscan.infrastructure.features.analysis.history.CodeBaseHistoryDTO;
import io.sch.historyscan.infrastructure.features.analysis.history.CodeBaseHistoryFileDTO;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class HistoryMapper {
    public CodeBaseHistoryDTO domainToWeb(CodeBaseHistory analyzedCodeBase) {
        return new CodeBaseHistoryDTO(
                analyzedCodeBase.commits().stream().map(this::domainToWeb).toList()
        );
    }

    public CodeBaseHistoryFileDTO domainToWeb(CodeBaseCommit codeBaseCommit) {
        return new CodeBaseHistoryFileDTO(
                this.domainToWeb(codeBaseCommit.info()),
                this.domainToWeb(codeBaseCommit.files())
        );
    }

    private List<CodeBaseHistoryCommitFileDTO> domainToWeb(List<CodeBaseHistoryCommitFile> files) {
        return files.stream().map(this::domainToWeb).toList();
    }

    private CodeBaseHistoryCommitFileDTO domainToWeb(CodeBaseHistoryCommitFile codeBaseHistoryCommitFile) {
        return new CodeBaseHistoryCommitFileDTO(
                codeBaseHistoryCommitFile.fileInfo().name(),
                codeBaseHistoryCommitFile.fileInfo().path(),
                codeBaseHistoryCommitFile.currentNbLines(),
                codeBaseHistoryCommitFile.nbAddedLines(),
                codeBaseHistoryCommitFile.nbDeletedLines(),
                codeBaseHistoryCommitFile.nbModifiedLines());
    }

    private CodeBaseHistoryCommitInfoDTO domainToWeb(CodeBaseHistoryCommitInfo info) {
        return new CodeBaseHistoryCommitInfoDTO(info.hash(), info.author(), info.date(), info.shortMessage());
    }
}

File: historyscan-back/src/main/kotlin/io/sch/historyscan/infrastructure/features/analysis/HistoryMapper.kt
package io.sch.historyscan.infrastructure.features.analysis

import io.sch.historyscan.domain.contexts.analysis.common.CodeBaseCommit
import io.sch.historyscan.domain.contexts.analysis.history.CodeBaseHistory
import io.sch.historyscan.domain.contexts.analysis.history.CodeBaseHistoryCommitFile
import io.sch.historyscan.domain.contexts.analysis.history.CodeBaseHistoryCommitInfo
import io.sch.historyscan.infrastructure.features.analysis.dto.history.CodeBaseHistoryCommitFileDTO
import io.sch.historyscan.infrastructure.features.analysis.dto.history.CodeBaseHistoryCommitInfoDTO
import io.sch.historyscan.infrastructure.features.analysis.dto.history.CodeBaseHistoryDTO
import io.sch.historyscan.infrastructure.features.analysis.dto.history.CodeBaseHistoryFileDTO
import org.springframework.stereotype.Component

@Component
class HistoryMapper {
    fun domainToWeb(analyzedCodeBase: CodeBaseHistory): CodeBaseHistoryDTO {
        return CodeBaseHistoryDTO(
                analyzedCodeBase.commits.stream().map { codeBaseCommit: CodeBaseCommit -> this.domainToWeb(codeBaseCommit) }.toList()
        )
    }

    fun domainToWeb(codeBaseCommit: CodeBaseCommit): CodeBaseHistoryFileDTO {
        return CodeBaseHistoryFileDTO(
                this.domainToWeb(codeBaseCommit.info),
                this.domainToWeb(codeBaseCommit.files)
        )
    }

    private fun domainToWeb(files: List<CodeBaseHistoryCommitFile>): List<CodeBaseHistoryCommitFileDTO> {
        return files.stream().map { codeBaseHistoryCommitFile: CodeBaseHistoryCommitFile -> this.domainToWeb(codeBaseHistoryCommitFile) }.toList()
    }

    private fun domainToWeb(codeBaseHistoryCommitFile: CodeBaseHistoryCommitFile): CodeBaseHistoryCommitFileDTO {
        return CodeBaseHistoryCommitFileDTO(
                codeBaseHistoryCommitFile.fileInfo.name,
                codeBaseHistoryCommitFile.fileInfo.path,
                codeBaseHistoryCommitFile.currentNbLines,
                codeBaseHistoryCommitFile.nbAddedLines,
                codeBaseHistoryCommitFile.nbDeletedLines,
                codeBaseHistoryCommitFile.nbModifiedLines)
    }

    private fun domainToWeb(info: CodeBaseHistoryCommitInfo): CodeBaseHistoryCommitInfoDTO {
        return CodeBaseHistoryCommitInfoDTO(info.hash, info.author, info.date, info.shortMessage)
    }
}
